### PR TITLE
SPI calculations

### PR DESCRIPTION
**Pull request by chrisjurich**
This PR contains the code needed to do substrate positioning index (SPI) calculations. This functionality is exposed through the function `enzy_htp.analysis.spi_metric()`, which is implemented in this PR. There is also an accompanying `structure_operation.closest_n_residues()` that helps to make spi calculations easier.

### CODE REVIEW

File: enzy_htp/structure/structure_operation/residue.py [hunk 1]
```diff
@@ -111,3 +111,34 @@ def check_res_topology_error(
...
+    Returns:
+        A List[Residue] containing the closest n residues.
+    """
+    target_coord = target.geom_center
+    distances = list()
+    for res in target.root().residues:
+        if res == target:
+            continue
+        
+        distances.append((mh.get_distance(target_coord, res.geom_center), res))
```
Comment (shaoqx) [inline enzy_htp/structure/structure_operation/residue.py:138]: Have mh being imported?

File: enzy_htp/structure/structure_operation/residue.py [hunk 2]
```diff
@@ -111,3 +111,34 @@ def check_res_topology_error(
     # 3. check for any bond (polyline) from the target residue thread through any ring
     # 4. check for any ring from the target residue is threaded by any bond
     pass
+
+
+def closest_n_residues(
```
Comment (shaoqx) [inline enzy_htp/structure/structure_operation/residue.py:116]: Please make a unit test for this func

File: enzy_htp/analysis/spi.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    result = list()
+    for ss in stru.structures:
+        result.append(
+            spi_metric( ss, ligand, pocket_sele )
+        )
+
+    return result
+
+
+
```
Comment (shaoqx) [inline enzy_htp/analysis/spi.py:87]: Please fix tha spacing

File: enzy_htp/analysis/spi.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        stru: The Structure() to analyze.
+        ligand: The Ligand() which will be the numerator in the spi metric.
+        pocket_sele: A pymol-formatted sele str which defines the denominator in the spi metric.
+
+    Returns:
+        A List[float] of len(StructureEnsemble.structures), with each value being a different spi metric for the nth Structure(). 
+    """
+
+    (lig_chain, lig_idx) = ligand.key()
+    session = interface.pymol.new_session()
```
Comment (shaoqx) [inline enzy_htp/analysis/spi.py:106]: Please use the context manager for pymol session here

File: enzy_htp/analysis/spi.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        result.append(
+            spi_metric( ss, ligand, pocket_sele )
+        )
+
+    return result
+
+
+
+@dispatch
+def spi_metric( stru: Structure,
```
Comment (shaoqx) [inline enzy_htp/analysis/spi.py:89]: Please make a unit test for this function using the result from old EnzyHTP (put there manually) as reference

File: enzy_htp/analysis/spi.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    (lig_chain, lig_idx) = ligand.key()
+    session = interface.pymol.new_session()
+    interface.pymol.load_enzy_htp_stru(session, stru )
+    results:List[Any] = interface.pymol.general_cmd(session,[
+        ('set', 'dot_solvent', 1),
+        ('create', 'ligand', f'chain {lig_chain} and resi {lig_idx}'),
+        ('create', 'protein', f'not (chain {lig_chain} and resi {lig_idx})'),
+        ('get_area', 'ligand'),
+        ('get_area', f'protein and ({pocket_sele})')
+    ])
```
Comment (shaoqx) [inline enzy_htp/analysis/spi.py:114]: It's better that we can put this in PyMolInterface in a function called get_spi() or a different name you like. (this makes it easier to maintain)

**Review by shaoqx — CHANGES_REQUESTED** [review]
Great PR! Love to see we have SPI in place. Please address the comments.

File: enzy_htp/structure/structure_operation/residue.py (hunk 1 above)
Comment (chrisjurich) [reply enzy_htp/structure/structure_operation/residue.py:138]: This is a good point! It is not yet.

File: enzy_htp/structure/structure_operation/residue.py (hunk 2 above)
Comment (chrisjurich) [reply enzy_htp/structure/structure_operation/residue.py:116]: Will do.

File: enzy_htp/analysis/spi.py (hunk 4 above)
Comment (chrisjurich) [reply enzy_htp/analysis/spi.py:106]: got it!

File: enzy_htp/structure/structure_operation/residue.py (hunk 1 above)
Comment (chrisjurich) [reply enzy_htp/structure/structure_operation/residue.py:138]: Updated and fixed this. Validated with unittests

File: enzy_htp/structure/structure_operation/residue.py (hunk 2 above)
Comment (chrisjurich) [reply enzy_htp/structure/structure_operation/residue.py:116]: Made a unit test!

File: enzy_htp/analysis/spi.py (hunk 3 above)
Comment (chrisjurich) [reply enzy_htp/analysis/spi.py:87]: fixed!

File: enzy_htp/analysis/spi.py (hunk 4 above)
Comment (chrisjurich) [reply enzy_htp/analysis/spi.py:106]: Added context manager

**Review by shaoqx — APPROVED** [review]
Look great! Thank you for this great new feature!

**Pull request merged**